File: tools/xgm2tool/src/sgdk/xgm2tool/format/GD3.java
package sgdk.xgm2tool.format;

import sgdk.tool.StringUtil;
import sgdk.xgm2tool.Launcher;
import sgdk.xgm2tool.tool.Util;

public class GD3
{
    final static String XGMTOOL_PRINT = "Optimized with XGMTool";

    int version;

    String trackName_EN;
    String trackName_JP;
    String gameName_EN;
    String gameName_JP;
    String systemName_EN;
    String systemName_JP;
    String authorName_EN;
    String authorName_JP;
    String date;
    String vgmConversionAuthor;
    String notes;

    public GD3()
    {
        trackName_EN = "";
        trackName_JP = "";
        gameName_EN = "";
        gameName_JP = "";
        systemName_EN = "";
        systemName_JP = "";
        authorName_EN = "";
        authorName_JP = "";
        date = "";
        vgmConversionAuthor = "";
        notes = "";

        version = 0x100;
    }

    public GD3(XD3 xd3)
    {
        trackName_EN = xd3.trackName;
        trackName_JP = xd3.trackName;
        gameName_EN = xd3.gameName;
        gameName_JP = xd3.gameName;
        systemName_EN = "SEGA Mega Drive";
        systemName_JP = "SEGA Mega Drive";
        authorName_EN = xd3.authorName;
        authorName_JP = xd3.authorName;
        date = xd3.date;
        vgmConversionAuthor = xd3.conversionAuthor;
        notes = xd3.notes;

        version = 0x100;
    }

    public GD3(byte[] data, int baseOffset)
    {
        int offset = baseOffset;

        if (!Launcher.silent)
            System.out.println("Parsing GD3...\n");

        final String id = Util.getASCIIString(data, offset, 4);

        if (!StringUtil.equals(id, "Gd3 "))
            System.out.println("Error: GD3 header not recognized !\n");
        offset += 4;

        // get version info
        version = Util.getInt32(data, offset);
        offset += 4;

        // get total size for GD3 data (we can ignore this field)
        Util.getInt32(data, offset);
        offset += 4;

        trackName_EN = Util.getWideString(data, offset);
        offset += (trackName_EN.length() + 1) * 2;
        trackName_JP = Util.getWideString(data, offset);
        offset += (trackName_JP.length() + 1) * 2;
        gameName_EN = Util.getWideString(data, offset);
        offset += (gameName_EN.length() + 1) * 2;
        gameName_JP = Util.getWideString(data, offset);
        offset += (gameName_JP.length() + 1) * 2;
        systemName_EN = Util.getWideString(data, offset);
        offset += (systemName_EN.length() + 1) * 2;
        systemName_JP = Util.getWideString(data, offset);
        offset += (systemName_JP.length() + 1) * 2;
        authorName_EN = Util.getWideString(data, offset);
        offset += (authorName_EN.length() + 1) * 2;
        authorName_JP = Util.getWideString(data, offset);
        offset += (authorName_JP.length() + 1) * 2;
        date = Util.getWideString(data, offset);
        offset += (date.length() + 1) * 2;
        vgmConversionAuthor = Util.getWideString(data, offset);
        offset += (vgmConversionAuthor.length() + 1) * 2;
        notes = Util.getWideString(data, offset);
        offset += (notes.length() + 1) * 2;

        // not ending with our signature ?
        if (!vgmConversionAuthor.endsWith(XGMTOOL_PRINT))
        {
            if (!StringUtil.isEmpty(vgmConversionAuthor))
                vgmConversionAuthor += " - ";

            // add XGMTool signature
            vgmConversionAuthor += XGMTOOL_PRINT;
        }
    }

    private int computeDataSize()
    {
        return (trackName_EN.length() * 2) + (trackName_JP.length() * 2) + (gameName_EN.length() * 2)
                + (gameName_JP.length() * 2) + (systemName_EN.length() * 2) + (systemName_JP.length() * 2)
                + (authorName_EN.length() * 2) + (authorName_JP.length() * 2) + (date.length() * 2)
                + (vgmConversionAuthor.length() * 2) + (notes.length() * 2) + (11 * 2);
    }

    public int getTotalDataSize()
    {
        return computeDataSize() + 12;
    }

    public byte[] asByteArray()
    {
        // align size on 2 bytes
        int size = computeDataSize();
        final byte[] result = new byte[size + 12];
        int offset = 0;

        // header
        System.arraycopy(Util.getBytes("Gd3 "), 0, result, offset, 4);
        offset += 4;
        // version
        Util.setInt32(result, offset, version);
        offset += 4;
        // size
        Util.setInt32(result, offset, size);
        offset += 4;

        // fields
        size = trackName_EN.length() * 2;
        System.arraycopy(Util.getBytes(trackName_EN, true), 0, result, offset, size);
        // +2 for 0 ending character in C
        offset += size + 2;
        size = trackName_JP.length() * 2;
        System.arraycopy(Util.getBytes(trackName_JP, true), 0, result, offset, size);
        offset += size + 2;
        size = gameName_EN.length() * 2;
        System.arraycopy(Util.getBytes(gameName_EN, true), 0, result, offset, size);
        offset += size + 2;
        size = gameName_JP.length() * 2;
        System.arraycopy(Util.getBytes(gameName_JP, true), 0, result, offset, size);
        offset += size + 2;
        size = systemName_EN.length() * 2;
        System.arraycopy(Util.getBytes(systemName_EN, true), 0, result, offset, size);
        offset += size + 2;
        size = systemName_JP.length() * 2;
        System.arraycopy(Util.getBytes(systemName_JP, true), 0, result, offset, size);
        offset += size + 2;
        size = authorName_EN.length() * 2;
        System.arraycopy(Util.getBytes(authorName_EN, true), 0, result, offset, size);
        offset += size + 2;
        size = authorName_JP.length() * 2;
        System.arraycopy(Util.getBytes(authorName_JP, true), 0, result, offset, size);
        offset += size + 2;
        size = date.length() * 2;
        System.arraycopy(Util.getBytes(date, true), 0, result, offset, size);
        offset += size + 2;
        size = vgmConversionAuthor.length() * 2;
        System.arraycopy(Util.getBytes(vgmConversionAuthor, true), 0, result, offset, size);
        offset += size + 2;
        size = notes.length() * 2;
        System.arraycopy(Util.getBytes(notes, true), 0, result, offset, size);

        return result;
    }
}

File: tools/crates/xgm2tool/src/gd3.rs
use crate::util;
use crate::xd3::XD3;

const XGMTOOL_PRINT: &str = "Optimized with XGMTool";

/// GD3 metadata tags (VGM format)
#[derive(Clone, Debug)]
pub struct GD3 {
    pub version: i32,
    pub track_name_en: String,
    pub track_name_jp: String,
    pub game_name_en: String,
    pub game_name_jp: String,
    pub system_name_en: String,
    pub system_name_jp: String,
    pub author_name_en: String,
    pub author_name_jp: String,
    pub date: String,
    pub vgm_conversion_author: String,
    pub notes: String,
}

impl GD3 {
    pub fn new() -> Self {
        GD3 {
            version: 0x100,
            track_name_en: String::new(), track_name_jp: String::new(),
            game_name_en: String::new(), game_name_jp: String::new(),
            system_name_en: String::new(), system_name_jp: String::new(),
            author_name_en: String::new(), author_name_jp: String::new(),
            date: String::new(), vgm_conversion_author: String::new(),
            notes: String::new(),
        }
    }

    pub fn from_xd3(xd3: &XD3) -> Self {
        GD3 {
            version: 0x100,
            track_name_en: xd3.track_name.clone(),
            track_name_jp: xd3.track_name.clone(),
            game_name_en: xd3.game_name.clone(),
            game_name_jp: xd3.game_name.clone(),
            system_name_en: "SEGA Mega Drive".to_string(),
            system_name_jp: "SEGA Mega Drive".to_string(),
            author_name_en: xd3.author_name.clone(),
            author_name_jp: xd3.author_name.clone(),
            date: xd3.date.clone(),
            vgm_conversion_author: xd3.conversion_author.clone(),
            notes: xd3.notes.clone(),
        }
    }

    pub fn from_data(data: &[u8], base_offset: usize) -> Self {
        let mut offset = base_offset;

        if !crate::is_silent() { println!("Parsing GD3..."); }

        let id = util::get_ascii_string_n(data, offset, 4);
        if id != "Gd3 " { println!("Error: GD3 header not recognized!"); }
        offset += 4;

        let version = util::get_i32(data, offset);
        offset += 4;
        // skip total size
        offset += 4;

        let track_name_en = util::get_wide_string(data, offset);
        offset += (track_name_en.len() + 1) * 2;
        let track_name_jp = util::get_wide_string(data, offset);
        offset += (track_name_jp.len() + 1) * 2;
        let game_name_en = util::get_wide_string(data, offset);
        offset += (game_name_en.len() + 1) * 2;
        let game_name_jp = util::get_wide_string(data, offset);
        offset += (game_name_jp.len() + 1) * 2;
        let system_name_en = util::get_wide_string(data, offset);
        offset += (system_name_en.len() + 1) * 2;
        let system_name_jp = util::get_wide_string(data, offset);
        offset += (system_name_jp.len() + 1) * 2;
        let author_name_en = util::get_wide_string(data, offset);
        offset += (author_name_en.len() + 1) * 2;
        let author_name_jp = util::get_wide_string(data, offset);
        offset += (author_name_jp.len() + 1) * 2;
        let date = util::get_wide_string(data, offset);
        offset += (date.len() + 1) * 2;
        let mut vgm_conversion_author = util::get_wide_string(data, offset);
        offset += (vgm_conversion_author.len() + 1) * 2;
        let notes = util::get_wide_string(data, offset);

        // Add XGMTool signature
        if !vgm_conversion_author.ends_with(XGMTOOL_PRINT) {
            if !vgm_conversion_author.is_empty() {
                vgm_conversion_author.push_str(" - ");
            }
            vgm_conversion_author.push_str(XGMTOOL_PRINT);
        }

        GD3 {
            version, track_name_en, track_name_jp,
            game_name_en, game_name_jp,
            system_name_en, system_name_jp,
            author_name_en, author_name_jp,
            date, vgm_conversion_author, notes,
        }
    }

    fn compute_data_size(&self) -> usize {
        (self.track_name_en.len() + self.track_name_jp.len()
            + self.game_name_en.len() + self.game_name_jp.len()
            + self.system_name_en.len() + self.system_name_jp.len()
            + self.author_name_en.len() + self.author_name_jp.len()
            + self.date.len() + self.vgm_conversion_author.len()
            + self.notes.len()) * 2 + (11 * 2)
    }

    pub fn get_total_data_size(&self) -> usize {
        self.compute_data_size() + 12
    }

    pub fn as_byte_array(&self) -> Vec<u8> {
        let data_size = self.compute_data_size();
        let mut result = vec![0u8; data_size + 12];
        let mut offset: usize = 0;

        // Header "Gd3 "
        result[0..4].copy_from_slice(b"Gd3 ");
        offset += 4;
        util::set_i32(&mut result, offset, self.version);
        offset += 4;
        util::set_i32(&mut result, offset, data_size as i32);
        offset += 4;

        // Helper to write wide string field
        let mut write_field = |s: &str| {
            let bytes = util::get_string_bytes(s, true);
            result[offset..offset + bytes.len()].copy_from_slice(&bytes);
            offset += bytes.len() + 2; // +2 for null terminator
        };

        write_field(&self.track_name_en);
        write_field(&self.track_name_jp);
        write_field(&self.game_name_en);
        write_field(&self.game_name_jp);
        write_field(&self.system_name_en);
        write_field(&self.system_name_jp);
        write_field(&self.author_name_en);
        write_field(&self.author_name_jp);
        write_field(&self.date);
        write_field(&self.vgm_conversion_author);
        write_field(&self.notes);

        result
    }
}
